### PR TITLE
Create NOTICE.txt

### PR DESCRIPTION
**Pull request by jasonblais**
Before submitting, please confirm you've
 - [X] read and understood our [Contributing Guidelines](https://github.com/mattermost/desktop/blob/master/CONTRIBUTING.md)
 - [X] completed [Mattermost Contributor Agreement](http://www.mattermost.org/mattermost-contributor-agreement/)
 - [x] executed `npm run lint:js` for proper code formatting

Please provide the following information:

**Summary**
Create NOTICE.txt

**Issue link**
Pre-requisite for inline spellchecker: https://github.com/mattermost/desktop/issues/225

**Test Cases**
N/A

**Additional Notes**
This is still a work-in-progress with a few additional steps
 - [x] @yuya-oc verify the source (github link) and description included for each dependency is correct
   - [x] @jasonblais update based on feedback
 - [x] @yuya-oc list any libraries that were modified. I wrote this document assuming none of them were
   - [x] @jasonblais properly attribute dependencies of `src/package.json` that were combined into a single JavaScript file by using webpack and hence modified (although the original dependencies weren't)
 - [x] @yuya-oc remove/replace the `chai-as-promised` dependency, since its license isn't acceptable: https://github.com/mattermost/desktop/pull/502
 - [x] @jasonblais confirm proper attribution of react libraries
 - [x] @jasonblais confirm proper attribution of babel libraries
 - [x] @jasonblais confirm proper attribution of electron-builder libraries
 - [x] @jasonblais add [`url-loaded` package](https://www.npmjs.com/package/url-loader)
 - [ ] @jasonblais ask for final review from it33

### CODE REVIEW

**Comment by yuya-oc** [comment]
Thanks, I will check.

Not urgent, but we should automate this process to maintain `NOTICE.txt` in future. Would you explain how to write the file?

> list any libraries that were modified. I wrote this document assuming none of them were

Now all dependencies are used as is. But dependencies of `src/package.json` are combined into a single JavaScript file by using webpack. This means original files are modified though, what about this?

**Comment by jasonblais** [comment]
> Not urgent, but we should automate this process to maintain NOTICE.txt in future. Would you explain how to write the file?

I just wrote it manually --

I don't think we need to automate it as we only need to update NOTICE.txt when we add/remove a dependency.

> Now all dependencies are used as is. But dependencies of src/package.json are combined into a single JavaScript file by using webpack. This means original files are modified though, what about this?

Awesome, that's useful to know. I added a follow-up task to make the necessary modifications.

File: NOTICE.txt [hunk 1]
```diff
@@ -0,0 +1,1052 @@
...
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+
+---
+
+## eslint-plugin-react
+
+This product contains 'eslint-plugin-react', a tool for React specific linting rules for ESLint, by Yannick Croissant.
```
Comment (yuya-oc) [inline NOTICE.txt]: ~a tool for~~ React specific linting rules for ESLint

File: NOTICE.txt [hunk 2]
```diff
@@ -0,0 +1,1052 @@
...
+IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+---
+
+## webpack-dev-server
+
+This product contains 'webpack-dev-server', which provides fast in-memory access to the webpack asset, by Webpack.
```
Comment (yuya-oc) [inline NOTICE.txt]: which serves a webpack app

**Comment by yuya-oc** [comment]
> I don't think we need to automate it as we only need to update NOTICE.txt when we add/remove a dependency, which happens quite rarely.

I understood by reading NOTICE.txt. It seems hard to automate.

**Comment by jasonblais** [comment]
Thanks! Updated with the following two commits:
 - https://github.com/mattermost/desktop/pull/488/commits/acedeb4f94de8c51d785c053180efc72fa93fe76
 - https://github.com/mattermost/desktop/pull/488/commits/7baf5c9542ce5874e5b18a6a120cce3f1570d012

**Comment by jasonblais** [comment]
In your comment here: https://github.com/mattermost/desktop/pull/488#issuecomment-289212099

> Now all dependencies are used as is. But dependencies of src/package.json are combined into a single JavaScript file by using webpack. This means original files are modified though, what about this?

Was the only modification that they were combined into a single JavaScript file, but otherwise no change?

File: NOTICE.txt [hunk 3]
```diff
@@ -1050,3 +1050,120 @@ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+---
+
+## babel
+
+This product contains the following packages of 'babel', a compiler for writing next generation JavaScript: 'babel-core', a Babel compiler core, 'babel-plugin-transform-object-rest-spread', a plugin that allows Babel to transform rest properties for object destructuring assignment and spread properties for object literals, and 'babel-preset-react', a Babel preset for all React plugins.
+'react-addons-css-transition-group', by Babel.
```
Comment (yuya-oc) [inline NOTICE.txt]: `react-addons-css-transition-group` is an addon of React. It's already written at React section.

File: NOTICE.txt [hunk 4]
```diff
@@ -1167,3 +1167,35 @@ NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
 LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
 WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+---
+
+## url-loader
+
+This product contains the following packages of 'url-loader', a url loader module for webpack, by Webpack.
```
Comment (yuya-oc) [inline NOTICE.txt]: by Webpack community. (see https://github.com/webpack-contrib)

File: NOTICE.txt [hunk 5]
```diff
@@ -0,0 +1,1201 @@
...
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+
+---
+
+## underscore
+
+This product contains 'underscore', a utility-belt library for JavaScript that provides support for usual functional suspects
+without extending any core JavaScript objects, by Jeremy Ashkenas, DocumentCloud and Investigative.
```
Comment (yuya-oc) [inline NOTICE.txt]: Probably "Jeremy Ashkenas, DocumentCloud and Investigative Reporters & Editors".

File: NOTICE.txt [hunk 6]
```diff
@@ -0,0 +1,1201 @@
...
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+---
+
+## url-loader
+
+This product contains the following packages of 'url-loader', a url loader module for webpack, by Webpack.
```
Comment (yuya-oc) [inline NOTICE.txt]: This product contains 'url-loader', ...

File: NOTICE.txt [hunk 7]
```diff
@@ -0,0 +1,1201 @@
...
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+---
+
+## babel-loader
+
+This product contains the following packages of 'babel-loader', a webpack plugin for Babel, by Babel.
```
Comment (yuya-oc) [inline NOTICE.txt]: This product contains 'babel-loader', ...

File: NOTICE.txt [hunk 8]
```diff
@@ -0,0 +1,1201 @@
...
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+---
+
+## babel-eslint
+
+This product contains the following packages of 'babel-eslint', an ESLint using Babel as the parser, by Babel.
```
Comment (yuya-oc) [inline NOTICE.txt]: This product contains 'babel-eslint', ...

**Comment by yuya-oc** [comment]
> Was the only modification that they were combined into a single JavaScript file, but otherwise no change?

I believe yes.

Just for note. If #387 was applied, combined files are definitely modified. On the other hand, `simple-spellchecker` is never modified because it's not combined into a single JS. It's not adapted to webpack (see [#277](https://github.com/mattermost/desktop/pull/277#issue-176242982)).

File: NOTICE.txt (hunk 3 above)
Comment (jasonblais) [inline NOTICE.txt]: sorry, copy-paste error :/

File: NOTICE.txt (hunk 8 above)
Comment (jasonblais) [inline NOTICE.txt]: and here too.

**Comment by jasonblais** [comment]
Fixed the typos and added a `contains a modified portion of...` for libraries in `src/package.json`. 

I have a task for myself to create a separate attribution list on how they were modified (were combined into a single JavaScript file by using webpack and hence modified. Plus perhaps further changes later with #387), but this will be sufficient for now and is how it's handled in other mattermost repos.

**Comment by jasonblais** [comment]
Let me know if this looks okay @yuya-oc and I can share with Ian for final review.

**Comment by jasonblais** [comment]
Awesome! I've queued it for my next meeting with Ian on Thursday

**Comment by jasonblais** [comment]
Approved by Ian

**Pull request merged**